Components EffectUnit rework

Be-ing opened the pull request:
This is a work in progress to rethink the EffectUnit UX for components. I'll be experimenting with different ways of focusing effects and showing that on the controller.

For now, the first commit just fixes a bug wherein the EffectUnit wasn't unfocused when show_parameters was set to 0.

Merge requirements:
  * [x] #1218 
  * [x] #1235

Be-ing committed:
Components: unfocus EffectUnit when show_parameters is set to 0

Be-ing commented:
The issue I am trying to address is that there is no visual indication on the controller of whether an effect is focused, and if so, which effect. Presently, the user must look at the skin to see this information, so it is easy to get confused about what turning the effect knobs will do. Here's a rough idea of how to map this differently:
  * Change name of showParametersButton to effectFocusButton.
  * When effectFocusButton is pressed and show_parameters is 0, set show_parameters to 1.
  * When show_parameters is 1, holding effectFocusButton and pressing an enableButton focuses that effect.
  * When effectFocusButton is held down and show_parameters is 1, light the LED of the effectFocusButton indicating the focused effect. When effectFocusButton is released, revert the effectFocusButton LEDs to indicate the enabled state of the effects.
  * While holding effectFocusButton, pressing the focused effectEnableButton resets show_focus to 0, but keeps show_parameters as 1.
  * When any effect is focused, light the LED of effectFocusButton to provide a visual indication on the controller that the knobs are controlling individual parameters and not metaknobs. When show_parameters is 1 (regardless of whether an effect is focused) press shift+effectFocusButton to unfocus and set show_parameters back to 0.

The biggest issue I imagine with this is that pressing effectFocusButton behaves differently depending on the state of show_parameters. If a user presses effectFocusButton and sees the effect unit expand in the skin, the logical assumption of how to go back to the previous collapsed state would be to press the same button again. Having to press shift+effectFocusButton to go back to the previous state is a discoverability issue.

Suggestions welcome.

Be-ing commented:
> The biggest issue I imagine with this is that pressing effectFocusButton behaves differently depending on the state of show_parameters. If a user presses effectFocusButton and sees the effect unit expand in the skin, the logical assumption of how to go back to the previous collapsed state would be to press the same button again. Having to press shift+effectFocusButton to go back to the previous state is a discoverability issue.

Perhaps alternate actions for short and long presses of effectFocusButton could overcome this:
  * When show_parameters is 1, short pressing effectFocusButton resets both focused_effect and show_parameters to 0.
  * When show_parameters is 1, holding effectFocusButton switches to the mode for choosing the focused effect as described in the previous comment.

ronso0 commented:
I just got used to Component JS for FX units and I like it. There's show_parameters, with Shift there's un/focus. And if -in minimal view- it's also possible to focus an effect without a forced expansion of the unit, that's perfect!
> Presently, the user must look at the skin to see this information, so it is easy to get confused about what turning the effect knobs will do.

Indeed, and there's no timed behaviour for physical buttons yet like there is for the virtual ones (long-press & release enables effect temporarily).
> Perhaps alternate actions for short and long presses of effectFocusButton

So, if timers enter the game, it could be sufficient to also apply them to LEDs, maybe an unobtrusive 'negative blink' like this:
Effect enabled, not focused = LED is On
Effect enabled and focused = LED is On, goes of for .1sec every second
I think, having to press a button to know which effect is focused, is equally intricate as looking at the screen. As I mentioned, I like how focus, highlight & show_parameters currently work. Because it's simple. Visual feedback about focus would make it pretty solid, I'd say.
> When show_parameters is 1, short pressing effectFocusButton resets both focused_effect and show_parameters to 0.

Make it a long-press and you got my vote :)

Be-ing commented:
> And if -in minimal view- it's also possible to focus an effect without a forced expansion of the unit, that's perfect!

Why do you want this? I think it would be confusing because the values the knobs would be controlling are not on screen.

> negative blink

I think blinking is annoying and I will not be implementing it.

> I think, having to press a button to know which effect is focused, is equally intricate as looking at the screen.

True, but presently there is no indication whether or not an effect is focused, so when looking at the controller, it's a guessing game as to what turning the knobs will do.

> Make it a long-press and you got my vote :)

That would make it too easy to accidentally unfocus and switch to collapsed mode. With your suggestion, if effectFocusButton was pressed to display the focused effect on the controller buttons' LEDs, there would be a very short window of opportunity (probably around a few hundred milliseconds) to let go of effectFocusButton without an unintended side effect.

ronso0 commented:
> Why do you want this? I think it would be confusing because the values the knobs would be controlling are not on screen.

Because it should be up to the user if he/she wants exanded view. For example, I got used to control Echo (delay & feedback independently) and Filter (HPF & resonance independently) blindly and I wouldn't want that expanded FX units squeeze my playlist.

> I think blinking is annoying and I will not be implementing it.

I think so too, but I would accept here if there's a subtle solution not a "LOOK AT ME" flash, if you know what I mean. Something like a pulse

> That would make it too easy to accidentally unfocus and switch to collapsed mode. With your suggestion, if effectFocusButton was pressed to display the focused effect on the controller buttons' LEDs

Hm, I meant short-press to mini/maximize like now, long-press to minimize and unfocus.

In general I feel, you already have a very precise idea, and that complex behaviour -if it would be fixed- would cut other options, other users' personal preferences. No offence, I very much respect your commitment on this issue

ronso0 commented:
Ẁhile your working on this anyway, I have another idea/wish/hint:
Instead of a shifted enable button cycling through the effects list in one direction in minimal view, shifted knobs could do that way better. I had this (and timers) in my script solution (low-level code admittedly, but only multiple timers for long-press would interfere now&then).
I decreased the shifted knobs' resolution so they would change to next/prev effect every +-10 steps of their actual value. Was pretty comfortable.

Be-ing commented:
That's a good idea. I'll keep it in mind for after the focusing rework.

Be-ing commented:
> Hm, I meant short-press to mini/maximize like now, long-press to minimize and unfocus.

Yeah, that's what I was thinking. Also I think long pressing effectFocusButton when show_parameters is 0 should both set show_parameters to 1 and switch the buttons to focus selection mode. That way, focusing an effect from collapsed mode will only require one long button press instead of one button press then another long button press.

ronso0 commented:
> Also I think long pressing effectFocusButton when show_parameters is 0 should both set show_parameters to 1 and switch the buttons to focus selection mode. That way, focusing an effect from collapsed mode will only require one long button press instead of one button press then another long button press.

In addition to that: while show_focus=1, enabling an effect could automatically focus it.

Be-ing commented:
> In addition to that: while show_focus=1, enabling an effect could automatically focus it.

I disagree. Enabling an effect and focusing should remain separate.

Be-ing committed:
Components: add new EffectAssignmentButton Component that toggles with decks

following Deere redesign to put effect unit assignment switches in decks
instead of effect units

also don't premake EffectUnit.enableOnChannelButtons

Be-ing committed:
Components: make EffectUnit togglable between effect units

Be-ing committed:
Components: rework EffectUnit UX

Replace showParametersButton with focusEffectButton. This button's LED now indicates whether
an effect is focused.
* Short press, parameters hidden: show parameters
* Short press, parameters showing: unfocus and hide parameters. The focused effect is remembered and restored
  the next time parameters are shown.
* Long press, parameters hidden or showing: switch to effect focus choose mode

In effect focus choose mode, the effect enable buttons act to select the focused effect. Their LEDs
show whether that effect is focused. When focusEffectButton is released, the effect buttons go back
to controlling the enable switches and their LEDs are updated to reflect which effects are enabled.

Be-ing commented:
Okay, I have implemented the focusing changes discussed above. I also added a new aspect to it. When hiding parameters, show_focus is reset to 0, but its value is stored and the focused effect is refocused when parameters are shown again. This seems to work as planned, but I haven't yet tried to actually mix with it. Please test.

I have removed the premade enableOnChannelButtons. They can still be made easily with `enableOnChannelButtons.addButton('CHANNEL_NAME')`, but I think they should only be used if the assignment buttons are in the effect unit like the Denon MC6000Mk2 (ping @uklotzde). Otherwise, I have added a new type of Button called EffectAssignmentButton that is separate from the EffectUnit object. The options object to the constructor of this requires a `group` option as the deck to enable and `effectUnit` as the number of the effect unit. I added a special case to Deck.setCurrentDeck so it will switch the deck of any EffectAssignmentButtons in the deck.

Be-ing commented:
Wow this is really fun! :D

uklotzde commented:
This is one of my tasks for 2.1, remap the effects unit for the MC6000MK2. Thanks for reminder ;)

Be-ing commented:
I have temporarily removed effect selection from this. I'll add it back to shift + knob turn after making a ControlObject to map to this.

Be-ing committed:
Components: use shift + effectFocusButton to toggle EffectUnits

Be-ing commented:
After discussing with @MelGrubb how to map the M-Audio Xponent, I realized there is some desire to switch physical EffectUnits on a controller between Mixxx's EffectUnits. I have implemented support for this as shift + effectFocusButton. To use it, pass an array of effect unit numbers to toggle between to the EffectUnit constructor, just like toggling between Decks: `new components.EffectUnit([1, 3]);`

Be-ing commented:
I have updated the [wiki](http://mixxx.org/wiki/doku.php/components_js#effectunit) to explain the new behavior and how to map it.

daschuer commented:
Thank you for working on this. The plain text on the wiki is hard to understand. Could you replace the term briefly by short and mention the 275 ms, this is more obvious for a non native speaker. 
A kind of truth table would also be very helpful. 

Is it now still possible to focus an effect without showing the buttons in the GUI? 
I think not, because a short press already sets show_parameters = 1. 
I think this is a valid requirement, because in a controller only scenario, all GUI space should be 
available for the library. 

Idea:
Would it more natural if we do it like that? 
* Hold effectFocusButton
* Short Press effectEanbleButton: Toggle the focus
* Long Press effectEnableButton: Toggle the focus and  show_parameters

Be-ing commented:
I will continue to work on the language on the wiki. It is complicated to explain, but it is actually quite intuitive to use.

> Is it now still possible to focus an effect without showing the buttons in the GUI?
I think not, because a short press already sets show_parameters = 1.
I think this is a valid requirement, because in a controller only scenario, all GUI space should be
available for the library.

As discussed above, no, it is not possible to have an effect focused without the parameters showing, both to save screen space and because it would be confusing to manipulate a focused effect without being able to see the values being manipulated.

> Would it more natural if we do it like that?
    Hold effectFocusButton
    Short Press effectEanbleButton: Toggle the focus
    Long Press effectEnableButton: Toggle the focus and show_parameters

I don't think so. The way it is currently, I can see whether an effect is focused from the LED of the effectFocusButton being lit. If I want to unfocus, it only requires one short press of the effectFocusButton.

ronso0 commented:
> Is it now still possible to focus an effect without showing the buttons in the GUI?
I think not, because a short press already sets show_parameters = 1.
I think this is a valid requirement, because in a controller only scenario, all GUI space should be
available for the library.

As I explained above, I'd like this to be possible, too (for me Filter & Echo are easy to use without GUI knobs). I understand that you try to build this for an expected majority of users for whom it's probably better to keep the interface simple/intuitive.
I would be glad if you leave easy to find comments in the code where to manually revert that automatic unfocussing behaviour and the knob behaviour for onHideParameters.

Be-ing commented:
Sure, I could even make a boolean argument to the EffectUnit constructor to control that behavior.

ronso0 commented:
That would be great!

Be-ing committed:
Components: add option to EffectUnit to allow focus without parameters showing

per @ronso0's request

Be-ing committed:
Components: fix soft takeover for EffectUnit knobs

Be-ing committed:
Components: provide common longPressTimeout on Button prototype

Be-ing commented:
Pass `true` as the second argument to the constructor to allow effects to be focused when parameters are hidden. You can easily make this a user configurable option for your script (when we have a UI for mapping options it will be even better).

IMO the biggest downside to using this option is that unfocusing requires too much thinking. You have to hold down the effectFocusButton and press the button for the focused effect to get back to controlling the metaknobs. With the default behavior you just have to short press the effectFocusButton when it its LED is lit.

Be-ing commented:
I have also moved the magic number for the timeout onto the prototype Button object instead of hardcoding it. If you want to experiment with different timeouts, set `components.Button.prototype.longPressTimeout`.

Be-ing commented:
By the way, I really fun chain I have been testing with is Bitcrusher -> Moog Filter -> Echo.

ronso0 commented:
> if you could test both ways that would be appreciated.

Yep, I will try this, as well as switching EffectUnits when switching decks. Sounds great!

Be-ing commented:
> Yep, I will try this, as well as switching EffectUnits when switching decks.

I only recommend toggling the EffectUnits with the Decks if you can't find a way to squeeze EffectAssignmentButtons into your mapping. It is helpful to keep different chains set up and swap them between decks during a set instead of tying each EffectUnit to a deck.

> Sounds great!

It does! And it's really fun!

Be-ing committed:
Components code cleanup and comments

MelGrubb commented:
If I could throw in one suggestion. Perhaps unfocusing the unit when the shift button is released would be more intuitive. The unshifted effect buttons should always toggle their corresponding EffectUnit. Shift+EffectButton could focus that effect. Pressing it multiple times could cycle the focus between Effects within that unit, and then when you release the shift, everything reverts to normal as with other shifted functions. Also, it opens up the possibility of using the button lights to indicate which effect units are active in normal mode. Then, when I press the shift button, the lights would change to show which unit has the focus, and then they would resume their enabled/disabled duties when shift is released.

 I'm going to try that myself in my Xponent mapping and see how it behaves.

Be-ing commented:
I think having multiple levels of focusing would be way too complicated to operate live. Assigning decks to different EffectUnits accomplishes effectively the same thing. With the typical layout of two effects sections on a controller, if you toggle between 1/3 on one side and 2/4 on the other side with shift + effectFocusButton you have access to all 4 of Mixxx's EffectUnits.

ronso0 commented:
I just tried this with a clean profile/mapping. So far I like very much:
- Changing focus & un/focusing is way easier now, just need one hand!
- Switching between effect units works well!

But knob behaviour is somehow screwed when focus is set (no matter which effect):
Knob1 > Meta1
Knob2 > Meta2
Knob3 > Parameter3 of focused effect
When focus is not set, knobs control Meta knobs correctly

I instantiated with default settings like this:
```
/* TerminalMix.effectUnit13 = new components.EffectUnit([1,3],true); */
TerminalMix.effectUnit13 = new components.EffectUnit([1,3]);
TerminalMix.effectUnit13.enableButtons[1].midi = [0x90, 0x07];
TerminalMix.effectUnit13.enableButtons[2].midi = [0x90, 0x08];
TerminalMix.effectUnit13.enableButtons[3].midi = [0x90, 0x09];
TerminalMix.effectUnit13.knobs[1].midi = [0xB0, 0x01];
TerminalMix.effectUnit13.knobs[2].midi = [0xB0, 0x02];
TerminalMix.effectUnit13.knobs[3].midi = [0xB0, 0x03];
/* still mapped like before (0.001-1) due to bug
TerminalMix.effectUnit13.dryWetKnob.midi = [0xB0, 0x2B]; */
TerminalMix.effectUnit13.effectFocusButton.midi = [0x90, 0x0A];
TerminalMix.effectUnit13.init();
```

Did I overlook another change?

Be-ing commented:
Hmm, that's weird. I can reproduce that. I must have broken something in the last commit. Looking into it.

Be-ing commented:
Oooof, I found the problem. It's not a bug in the JavaScript. It's the bug in `engine.connectControl` that I fixed in #1218 :facepalm: You need to compile with #1218 for this to work.

Be-ing commented on res/controllers/midi-components-0.0.js:
```diff
@@ -583,11 +594,16 @@
         };
         this.EffectUnitKnob.prototype = new Pot({
             group: this.group,
+            input: function (channel, control, value, status, group) {
```
If you comment out this function it'll work without compiling #1218. It'll use Pot.prototype.input() instead.

ronso0 replied:
Ha, just deleted build and started scons...

Be-ing replied:
@MelGrubb if you want to try this with a master build from the build server before #1218 is merged you'll need to comment out this function.

ronso0 commented:
Okay, I compiled with #1218 and set it set up like before
```TerminalMix.effectUnit13 = new components.EffectUnit([1,3]);```
- knobs are fine!
- switching between units works!

But with a clean mixxx.cfg there are no focus buttons. If I add this to controller script for units 1-4
```engine.setValue("[EffectRack1_EffectUnit1]","show_focus",1);```
focus buttons are there. They are still there after next start, even if I remove those lines from controller script or unplug controller. Seems "show_focus" is persistent, don't know if this is how intended it to be..
But this doesn't bother me.

Setting effect units up like this (without "show_focus" in controller script)
```TerminalMix.effectUnit13 = new components.EffectUnit([1,3],true);```
would show focus buttons only for 1st unit in array, 1 in my case.

Hope this helps!
Will give it an intense try in a set this night...but feels very nice so far!!

Be-ing committed:
Components EffectUnit refactor to solve some bugs

Move some initialization code into EffectUnit.setCurrentUnit and call
setCurrentUnit on construction to ensure that state of show_focus and
show_parameters are set correctly when toggling units. Also refactor
code for showing/hiding focus controls in skin to a callback connected
to show_parameters so it gets triggered by clicking the show_parameters
button in skins as well as effectFocusButton on the controller.

Be-ing commented:
Thanks for testing. I fixed issues with setting show_focus on startup and when toggling between EffectUnits. Also fixed an issue with soft takeover not getting activated when toggling. And I made toggling the show_parameters button in skins trigger the same actions as when show_parameters is toggled by pressing/releasing effectFocusButton to keep everything in sync.

I am wondering if maybe show_focus should always be 1, even if show_parameters is 0 in the default mode (not passing `true` as the second argument to the constructor). When passing `true` as the second arg to the constructor, the focus buttons in the skin have the nice side effect of showing which unit the controller is manipulating. Letting the buttons be shown on screen when show_parameters is 0 would allow manipulating the individual parameters without taking up the screen space when show_parameters is 1 by clicking the buttons on screen. However, that would not be accessible from the controller and if you wanted to see the parameters, the effect would be unfocused when hiding them again.

> Seems "show_focus" is persistent, don't know if this is how you intended it to be..

It is, and that was intentional, but now I am doubting that it should be. If a user unplugs a controller with a mapping that had set show_focus to 1 and restarts Mixxx, the focus should not be shown because it won't do anything.

Be-ing committed:
Components: add selecting next effect back to shift + enableButton

Be-ing commented:
@ronso0, any more feedback?

ronso0 commented:
It's a bit hard to test since I noticed my Focus buttons have defective contacts. Apart from that it works great to un/focus with just one hand. I can try with recent changes tomorrow.

> When passing true as the second arg to the constructor, the focus buttons in the skin have the nice side effect of showing which unit the controller is manipulating.

Indeed! This way there's no need for an additional indicator (in skin).

Do you plan to map effect selection to a shifted knob in this PR? Shifted button takes too long I think, I noticed I'd go back to skin to change effects.

Be-ing commented:
> It's a bit hard to test since I noticed my Focus buttons have defective contacts.

Dang, that sucks. You could try opening your controller and cleaning the contact and the underside of the pads.

> Do you plan to map effect selection to a shifted knob in this PR? Shifted button takes too long I think, I noticed I'd go back to skin to change effects.

Initially I was thinking of making a new ControlPotmeter for this so if the user knew specifically what effect they wanted to switch to, they could move the knob to a specific position. However, I realized this would be awkward because the position of the ControlPotmeter would not be shown on screen. It would also take quite a bit of work to implement. I'll try using effect_selector with the relative change in the position of the knob and divide the range of the knob by the number of available effects so any effect could be selected with one turn of the knob.

ronso0 commented:
> Dang, that sucks. You could try opening your controller and cleaning the contact and the underside of the pads.

I opened my controller often enough ;) Actually I replaced all 56 worn-out micro push buttons recently because the controller (layout & handling) is worth it. Cost me 10 € and every buttons works fine except that one...

> I'll divide the range of the knob by the number of available effects so any effect could be selected with one turn of the knob. I think I'll need to implement the number of available effects as a new read-only CO. For now, I can hard code it, but that won't be future proof.

Sounds like a lot of work. My experience with my previous implementation of this approach: on my controller the knobs have an responsive range of ~250degrees (values 0-127) and switched effects every 6 steps (~12degrees), which would feel good. My rotary encoders have 13 sensible clicks in that range.

ronso0 commented:
If users have own LV2 effects it might get difficult with the range divided by effects.

Be-ing commented:
> Sounds like a lot of work.

I think it will be pretty easy.

> If users have own LV2 effects it might get difficult with the range divided by effects.

Yes, but what if a user wants to select an LV2 effect?

ronso0 commented:
I think we talk about different things: if I understand you correctly, you want to map effect regions onto the knob's range, with the sensitivity problem that arises with more than let's say 20 effects.
I imagine another issue. Let's say the effect knob in question is at 2 o'clock and I currently have Echo selected. I want to have Filter. In relative mode this would require just a little turn clock-wise since those two effects are neighbours in the effect list.
In absolute mode, turning the shifted knob just a little would make the effect slot jump to a completely different effect. There I'd have to think in which direction to turn the knob, and how far approximately, to get my desired effect.
Those two methods only differ (besides the sensitivity problem) if effects are sorted alphabetically like in EffectSelector widget. Only then one can guess where they are located on a A-Z scale.

In my script I just take the first value the shifted knob sent (startValue), waited until it changed by +-5, then switched the effect with [effect_selector,+-1] (and reset startValue). If the end of the knob's physical range is reached before the desired effect is selected, just unshift, turn the knob back, press shift and continue turning. This was intuitive and worked quite well, like with a ratchet screwdriver.

Be-ing commented:
> Let's say the effect knob in question is at 2 o'clock and I currently have Echo selected. I want to have Filter. In relative mode this would require just a little turn clock-wise since those two effects are neighbours in the effect list. In absolute mode, turning the shifted knob just a little would make the effect slot jump to a completely different effect.

That could be alleviated with soft takeover, but it would be even more awkward than soft takeover normally is because the position of the effect selection knob wouldn't be displayed on screen. So yes, I agree that using the relative effect_selector CO will be better. The implementation will be something like #1227.

ronso0 commented:
Cool, let's see if it works out!

> The implementation will be something like #1227.

I'm sorry, I'm still struggling to visualize abstract code, I need 'pictures' to make it tangible. Could you describe how this behaves in softTakeover situation?

MelGrubb commented:
Could you use a different control entirely? For instance, if you're already in the "mode" to select from the available effects, could you take over the encoder that's normally used to browse the library? You're clearly in the middle of a different operation and wouldn't be browsing the library right now anyway, so reusing that control seems safe enough, and at least somewhat mnemonic. You just need to stop thinking of it as the library browser and start thinking of it as the "list" browser.

On Mar 28, 2017 1:40 AM, Be <notifications@github.com> wrote:

Let's say the effect knob in question is at 2 o'clock and I currently have Echo selected. I want to have Filter. In relative mode this would require just a little turn clock-wise since those two effects are neighbours in the effect list. In absolute mode, turning the shifted knob just a little would make the effect slot jump to a completely different effect.

That could be alleviated with soft takeover, but it would be even more awkward than soft takeover normally is because the position of the effect selection knob wouldn't be displayed on screen. So yes, I agree that using the relative effect_selector CO will be better. The implementation will be something like #1227<https://github.com/mixxxdj/mixxx/pull/1227>.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/mixxxdj/mixxx/pull/1211#issuecomment-289668850>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABGZB3_15YXVsH-06tPhfpidUiNHii9fks5rqJ1WgaJpZM4MXIZi>.

Be-ing commented:
> I'm sorry, I'm still struggling to visualize abstract code, I need 'pictures' to make it tangible. Could you describe how this behaves in softTakeover situation?

The point of using a relative control would be that soft takeover is not necessary, as you described above.

> Could you use a different control entirely? For instance, if you're already in the "mode" to select from the available effects, could you take over the encoder that's normally used to browse the library? You're clearly in the middle of a different operation and wouldn't be browsing the library right now anyway, so reusing that control seems safe enough, and at least somewhat mnemonic. You just need to stop thinking of it as the library browser and start thinking of it as the "list" browser.

No, because it wouldn't be clear which effect to change. The effect focus could be repurposed for that, but it wouldn't be as intuitive as pressing shift and turning the effect's metaknob.

ronso0 commented:
Hmkay, so it's just what I described but generalized as pot mode for any purpose.

> Could you use a different control entirely? For instance, if you're already in the "mode" to select from the available effects, could you take over the encoder that's normally used to browse the library?

That's a nice idea, encoder would be great for this purpose! But I agree that it wouldn't be intuitive, this encoder could be really far away from effect section, maybe not everybody has one.

ronso0 commented:
I tested new behaviour and it works fine!
Some small issues:
- Focus button LED indicates if an effect is focused, but after restart I'd have to un-focus and then re-focus an effect to make it work
- as I understood it, long-pressing Focus button should light the the focused effect's LED, but it doesn't

Again, it's nice that dis/appearing Focus buttons in effect units indicate focused unit when switchng between them!

Be-ing commented:
@ronso0 are the issues you encountered when passing `true` as the second parameter to the EffectUnit constructor, without that, or in either case?

ronso0 commented:
> are the issues you encountered when passing true as the second parameter to the EffectUnit constructor, without that, or in either case?

Both issues happen both with and without `true`.
To avoid any confusion, this is how I set up effect units to test it:
```
TerminalMix.effectUnit13 = new components.EffectUnit([1,3],true);
TerminalMix.effectUnit13.enableButtons[1].midi = [0x90, 0x07];
TerminalMix.effectUnit13.enableButtons[2].midi = [0x90, 0x08];
TerminalMix.effectUnit13.enableButtons[3].midi = [0x90, 0x09];
...knobs...
TerminalMix.effectUnit13.effectFocusButton.midi = [0x90, 0x0A];
TerminalMix.effectUnit13.init();
```
```
TerminalMix.effectUnit24 = new components.EffectUnit([2,4]);
TerminalMix.effectUnit24.enableButtons[1].midi = [0x91, 0x07];
TerminalMix.effectUnit24.enableButtons[2].midi = [0x91, 0x08];
TerminalMix.effectUnit24.enableButtons[3].midi = [0x91, 0x09];
...knobs...
TerminalMix.effectUnit24.effectFocusButton.midi = [0x91, 0x0A];
TerminalMix.effectUnit24.init();
```
Without `true` focus buttons are shown on both units [2,4] if they are both expanded...missing 'that nice side effect'
Hope it helps

Be-ing commented:
> Focus button LED indicates if an effect is focused, but after restart I'd have to un-focus and then re-focus an effect to make it work
> as I understood it, long-pressing Focus button should light the the focused effect's LED, but it doesn't

I cannot reproduce these issues. Try fetching and compiling latest master and test again.

ronso0 commented:
> I cannot reproduce these issues. Try fetching and compiling latest master and test again.

Okay, I compiled #1211, #1218 & #1151 and tested again.
Test A
effect 1: enabled
effect 2: disabled, focused
press FocusButton: no LED lights up

Test B
effect 1: disabled
effect 2: enabled, focused
press FocusButton: LED 2 lights up

So, appearantly longpressed FocusButton only lights focused effect's LED if it's also enabled.

____
FocusLED is still not in sync after starting mixxx:
On unit 2 I have effect 2 focused. I close & restart mixxx. Focused effect is restored, LED not.

Be-ing commented:
I encountered a strange bug yesterday. Occasionally, with parameters not showing, turning the knobs would not change the metaknobs. I don't think it was an issue with soft takeover because I could turn the physical knob across its entire range and nothing would happen. It only worked again when I pressed the effectFocusButton. Perhaps there was an issue setting the `inKey` and/or `group` properties for the knobs.

Be-ing committed:
Components: change effects in EffectUnit by shift + knob turn

and remove changing effects from shift + button press

Be-ing commented:
:/ It seems that loading initial knob positions on startup is not working anymore...

Be-ing committed:
Components: fix loading of initial values for EffectUnit knobs

Be-ing commented:
> FocusLED is still not in sync after starting mixxx:
 On unit 2 I have effect 2 focused. I close & restart mixxx. Focused effect is restored, LED not.

I still cannot reproduce this issue. You are referring to the effectFocusButton LED on startup, right? No enableButtons should not be lit on startup.

Please test compiling with #1235 and using the new shift + knob turn functionality to switch effects.

Be-ing commented:
I did some usability testing with a friend a few days ago. I confirmed that the focusing functionality is totally out of the way if the user doesn't know it is there. I think this is good, because the focusing functionality is complicated. It was difficult for him to figure out after I gave a hint to try long pressing the buttons. It's something that needs to be explained. I'll make a wiki page targeted at users that can be linked from the wiki page of each controller that uses this.

Be-ing committed:
Components: support 14 bit EffectUnitKnobs

Be-ing commented:
I think this is ready for merge. I have not been able to reproduce the issue @ronso0 reports, but it is a minor issue that can be fixed later if we figure out how to reproduce it. I'd like to get this and #1227 merged now so they can be used in #1233.

Be-ing committed:
Components: fix order of operations for 14 bit MIDI EffectUnitKnobs

Be-ing commented:
Ready for merge?

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into components_effectunit_rework

Be-ing commented:
Please merge. This is holding up controller mappings.

daschuer commented:
OK I will merge now. 
@ronso0: It would be nice to get a LGTM from you nevertheless to make this PR complete .. or if there are remaining issues a according comment. In the later case we can open a new PR or file a bug. Thank you.

daschuer commented:
Thank you for working on this.

ronso0 commented:
Yeah, the parts of this I've been using work great: effect toggling, focus toggle, parameter control, meta control, expand/collapse, switching between effect units. Though, I can't judge the code, at all.

I can test again next week to see if the bugs are still present (LED of focused effect only lights up if this effect is ON)
Once I experienced the bug @Be-ing described, about a (meta) knob losing contact to GUI knob. Caused an awkward moment during a live gig but when I noticed I could catch it with softTakeover.

For the Components wishlist: Temporarily de/activating an effect with longpress (like in GUI & like with previous effect mappping) would a nice feature, it's fun to use with beatlooproll!

Be-ing commented:
> For the Components wishlist: Temporarily de/activating an effect with longpress (like in GUI & like with previous effect mappping) would a nice feature, it's fun to use with beatlooproll!

Interesting idea. I'll play with that :)

Be-ing commented:
I wrote a [wiki page](https://mixxx.org/wiki/doku.php/standard_effects_mapping) targeted at users explaining how this works. Feel free to edit it if you think the language could be clearer.

Be-ing commented:
@ferranpujolcamins have you made any progress updating the Xone K2 mapping with this?

ferranpujolcamins commented:
No, I'm sorry. I won't be able to work on this until next month.

On 12 May 2017 5:41 a.m., "Be" <notifications@github.com> wrote:

> @ferranpujolcamins <https://github.com/ferranpujolcamins> have you made
> any progress updating the Xone K2 mapping with this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1211#issuecomment-300975991>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGIcT5yvAcooI2ytmPyP6AzWKpHE0Q0cks5r49TPgaJpZM4MXIZi>
> .
>

ronso0 commented:
Just tested: working great, like it did so far!
Thanks to your other fix, effect toggling while fully dry now works reliably.

The issue with unlit Focus button after restart is due to this snippet in controller script ([here](https://github.com/mixxxdj/mixxx/blob/master/res/controllers/Reloop%20Terminal%20Mix%202-4.js#L27)):
```
// Extinguish all LEDs
for (var i=0; i<=3; i++) {  // 4 decks
	for (var j=1; j<=120; j++) {
		midi.sendShortMsg(0x90+i,j,0x00); 
	}
}
```
So it's kinda solved, I guess.

The other issue I found persists...
Press Focus button > focused effects LED should light up > only does so if this effect is enabled
...but this is not a problem for me anymore since I got used to look at the screen for parameter adjustment due to softTakeover.

Be-ing commented:
> The other issue I found persists...
Press Focus button > focused effects LED should light up > only does so if this effect is enabled
...but this is not a problem for me anymore since I got used to look at the screen for parameter adjustment due to softTakeover.

Maybe you're misunderstanding how it is meant to work. Regardless of the state of show_parameters, you need to press and hold effectFocusButton to show which effect is focused. Short pressing effectFocusButton will toggle show_parameters.

Be-ing commented:
@ronso0 does commit fb79a04fa2adffd3888e29643c83fdad22d5547d from PR #1233 fix the issue you had with the LEDs showing the focused effect?

ronso0 commented:
@Be-ing Yes, now it works: focus led lights up no matter if effect is enbaled.
Thanks!

Be-ing commented:
Thank @DJMaxergy for fixing that.

Be-ing commented:
I am thinking of making the buttons control the first 3 button parameters when an effect is focused. When I was first working on this, I found it helpful to retain access to the enable buttons of all the effects in the chain when an effect is focused, but I haven't actually been making use of that much. #1256 made it clear to me that button parameters are actually useful and it would be nice to have access to them on controllers. What do you think? @ronso0, @DJMaxergy?

ronso0 commented:
I understand the Triplet example.
Shift + effect buttons could be mapped to mixxx effect enable buttons.
Because it's good to always have a reliable way to quickly switch effects on or off, without any modifier key. For example when something goes wrong (see 1). Also using the effect enable switch in power window mode together with loop roll buttons is very interesting > press both buttons simultaneously, optionally turn Meta knob

1) i.e. feedback goes through the ceiling with Echo or Reverb, or when Meta knob loses grip on a linked effect knob (happened to me twice with Moog filter and I can't tell if the reason was Components or the knob sending faulty values)

Be-ing commented:
Good point, having that access is helpful sometimes. How about when an effect is focused, the buttons control the focused effect's button parameters and the LEDs show the state of the button parameters. With shift and an effect focused, the buttons could be used to toggle the enable switches of all the effects in the chain without showing the state of the enable switches on the buttons' LEDs.

> Meta knob loses grip on a linked effect knob

Does this happen after loading an effect? If so, that's because of #1148.

ronso0 commented:
Hm, you mean the exact opposite of my proposal..
For my loop roll example you'd have to press three buttons at once then (while pressing Shift and loop roll buttons might do sometime completely different).
I admit using effect toggle + loop roll while an effect is focused seems to be a rare use case.
But more importantly I consider effect toggles to be basic controls like Play button and volume fader, which should work the same way under all circumstances, whereas IMO single effect parameter knobs and buttons are on a special level and could require modifiers like Shift.

Be-ing commented:
>  I admit using effect toggle + loop roll while an effect is focused seems to be a rare use case.

What about using effects with loop rolling would require the effect being focused? Why not use the enable switch together with a rolling loop without the effect focused?

Regardless of what the buttons are doing, the dry/wet knob is always there as an emergency brake.

ronso0 commented:
> Does this happen after loading an effect?

No, it happened somewhere during a session, the effect was in use since the beginning.
when turned quickly the physical knob sometimes sends values like 56-57-01-80-62-63, though IIRC the meta knob only lost the low pass knob.

ronso0 commented:
> What about using effects with loop rolling would require the effect being focused?

How to switch on an effect then?
I use loop roll very spontaneous. It's quite the opposite having to focus an effect in advance to enable that feature.

> Why not use the enable switch together with a rolling loop without the effect focused?

Sure, that's how I do in most cases, I wanted to point you to other valid use cases.
These are just my 2 cents..

DJMaxergy commented:
I agree with @ronso0:

> But more importantly I consider effect toggles to be basic controls like Play button and volume fader, which should work the same way under all circumstances, whereas IMO single effect parameter knobs and buttons are on a special level and could require modifiers like Shift.

Someone who gets in touch with a controller and it's effect buttons would be very confused, I think, if he cannot activate effects with the concerning buttons on the controller. It's a key feature at the currently available controllers. Yes, you can read the documentation and get into it, but it's not very intuitive. I sometimes forget or doesn't see immediately which effect I have focused and then I would wonder why I cannot activate the other effects.

Be-ing commented:
Please give #1353 a try. I am finding it more intuitive than the old way where the knobs controlled the focused effect and the buttons controlled the chain. Now both the knobs and the buttons control the focused effect.

@DJMaxergy, I have done usability testing with this and confirmed that it is unlikely that anyone would focus an effect on accident and get confused about what is going on. I intentionally designed the focus switching in a way that the user needs to know it is there to activate it.

The focus button's LED shows whether an effect is focused. If it is lit, you can know that touching the knobs (and with #1353, the buttons too) will control the focused effect. If you forget which effect you have focused, you can either look at the screen or hold down the focus button which will show which effect is focused on the buttons' LEDs until you release the focus button.

I should really make a video demonstrating all this.
